src/data/right_inv.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright © 2021 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import data.bundle
+import data.equiv.transfer_instance
+import data.pi
+
+/-!
+# Bundled right inverse (section)
+
+In this file we implement the bundled version of a right inverse function. The final purpose of
+this type is to conviniently implement sections of bundles.
+
+## Implementation notes
+
+Let `E : B → Type*` be the collection of fibers of a bundle on `B`.
+In the case of sections of bundles we prove that the type of right inverses of `bundle.proj E` is
+equivalent to the type `Π x : B, E x`. Both implementations of sections have their advantages:
+`right_inv (proj E)` is better to talk about properties such as continuity, whereas `Π x : B, E x`
+works better with algebraic structures: it naturally inherits the algebraic structures present on
+the fibers, thanks to the `Π` instances. It is hence a good idea to implement both and write down
+the equivalence between them, but then we have to choose only one for our final implementation of
+sections. We chose `right_inv (proj E)` because it is more general (some of the results in the
+theory of sections are true for `right_inv f`, for a generic `f`), and it has a more natural
+coercion to a function `B → bundle.total_space E` which is finally what we want. Note that we could
+also register a coercion to `B → bundle.total_space E` from `Π x : B, E x`, and it is not a bad idea
+to implement things this way, but it is just slightly more clumsy: in the case of continuous
+sections, in the first case one can implement continuous sections as
+`structure continuous_section (f : α → β) extends right_inv f, C(β, α)`
+whereas in the second way one would have to write continuity manually; it is not a big deal but the
+first way is slightly more natural. Also note that to recover the value of the section on the fiber
+in the first way one can use `.2` whereas the second way would make use of coercions.
+
+-/
+
+/-- Bundled right inverse of a function. Note that here the `nolint` has a true mathematical
+meaning: the structure is inhabited iff the function is surjective. -/
+@[nolint has_inhabited_instance]
+structure right_inv {α: Type*} {β: Type*} (f : α → β) :=
```

Comment:
I ended up doing something similar in #5829, but with `f` bundled.

Reply:
Yep but the two are different mathematically, yours is the type of function with an inverse, mine is with a given inverse, so there's no conflict here, but at most one would want to register a coercion from my type to yours... you agree?

Reply:
Indeed they are not substitutes for each other, but I thought it was worth noting the related development.